new/ohai_identity_attributes.md

```diff
@@ -0,0 +1,124 @@
+  it was made consistent across all platforms -- in some cases this was the return of `/bin/hostname` and
+  Solaris used hostname + NIS domain).
+* `domain`: computed from the domain suffix of the `fqdn` attribute
+
+Changing the `hostname` to be something other than the short hostname is most likely unwise since it has been
+that way consistently since it was introduced.  Because of that the `machinename` attribute was introduced to
+expose the actual return of `/bin/hostname` which is used in `node_name` guessing in the Chef::Client which
+uses the algorithm:
+
+```ruby
+name = Chef::Config[:node_name] || ohai[:fqdn] || ohai[:machinename] || ohai[:hostname]
+```
+
+It was discussed making the `fqdn` attribute to prefer the return of `/bin/hostname` when that returned a
+FQDN, but this felt to break some edge conditions in DNS related cookbooks (djbdns?) and so this was forced
+to be the fqdn which was resolved via DNS resolution across all platforms.  For platforms that do not support
+the `-f` flag to `hostname` a pure ruby version is used by ohai.
+
+The changes in Ohai 7.0.0 to the `fqdn` attribute broke Solaris people with busted DNS who expected the
+NIS domainname workaround to work.
+
+There is also a weakness in the `fqdn` attribute in that it relies on having setup forward and reverse DNS and that
+DNS is operating correctly.  A DNS failure may cause the fqdn attribute to be nil transiently.  This is likely
+a problem because Chef::Client uses it internally as first precedence for guessing the `node_name` and `fqdn` is
+used by many cookbooks.
+
+# Specification
+
+## First Proposal (Short Term Fix)
+
+The first proposal is to modify the `fqdn` attribute to be:
+
+# Equivalent to `/bin/hostname -f` if that algorithm succeeds in returning a FQDN
+# Returning `/bin/hostname` if that returns a FQDN
+# Returning `/bin/hostname` plus `/bin/domainname`
+
+This is not a strict revert back to the Ohai 6.x behavior and it should be implemented consistently across
+all platforms.  It will fix the Solaris behavior to construct the FQDN from the NIS domain in the absence of
+working DNS when there's only a short hostname.
+
+This will potentially break the semantics that the FQDN is supposed to be computed from the DNS domain.
+
+Embedding this information into a different attribute will not solve the problem because too many cookbooks
+expect the `fqdn` attribute to Just Work(tm) and there would be too many cookbooks to update.
+
+This changes the semantics of the `fqdn` attribute and may be a breaking change, but would need to be
+introduced in a minor release of Chef 12 in order to fix the issue for Solaris customers.  The risk is low.
+
+## Second Proposal (Long Term Fix)
+
+Since the short term fix breaks the strict semantics of the `fqdn` attribute, that opens the door as to if
+those semantics were right in the first place.
+
+The `hostname -f` lookup has an inherent weakness in that it relies on DNS, which may fail.  It is also obvious
```

Comment:
Yeah, I ran across a Stackexchange issue recently where someone said they'r hosts file had to be "ip_address alias canonical_hostname" for some other software, but Chef was using "alias" for the fqdn after the first run. Which of course, is expected behavior, and it's the other application that's doing it wrong.

---
The `hostname -f` looking has an inherent weakness in that it relies on the system's resolver. Depending on the resolver configuration, we may receive incorrect answers due to intentional misconfiguration of the hosts file, or even failures due to unavailable DNS, NIS, etc. It is also obvious that both cookbook users and authors expect the `fqdn` attribute to DWIM.  The use case for having the `fqdn` attribute be strict is a very small subset of a few cookbooks and is the uncommon case.